astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -653,6 +705,58 @@ def max(self, **kwargs):
             np.max, propagate_uncertainties=propagate_uncertainties, **kwargs
         )
 
+    @sharedmethod
+    def _convert_operand_class_with_suitable_dtype(self, cls, operand, ref=None):
+        """
+        Convert an operand to the specified class, first converting Python
+        scalars to a NumPy dtype that preserves their expected casting behaviour
+        under arithmetic with a reference operand or dtype (see NEP 50).
+
+        Parameters
+        ----------
+        cls : class
+            Class to create an instance of, with ``operand`` as the argument.
+
+        operand : object convertible to ``cls`` (`~astropy.nddata.NDData`-like)
+
+        ref : ``NDData``-like, `~numpy.ndarray`-like or `~numpy.dtype`, optional
+            Reference operand/dtype with which arithmetic is to be performed.
+
+        Returns
+        -------
+        result : instance of ``cls``
+        """
+        # Python scalars receive special "weak typing" treatment in NumPy 2
+        # (NEP 50) and need to be cast explicitly to the appropriate dtype
+        # before conversion to NDData, otherwise the instance just defaults to
+        # 64 bits, leading to unwanted upcasting in subsequent arithmetic,
+        # compared with NumPy. This weak typing does not apply to *lists* of
+        # Python scalars etc., which NumPy converts to arrays in the default
+        # way when doing arithmetic. The "__getitem__" check below excludes
+        # "scalars" that are NumPy types, strings or buffers.
+        if (ref is not None and np.isscalar(operand) and
+            not hasattr(operand, '__getitem__')):
+
```

Comment:
Most of the logic here is to handle odd cases where the arithmetic method is called from the instance or class with 2 arguments (_other_ than `self`), in which case the _first_ operand _could_ be a scalar and we need to compare its type with operand2, which can be anything. In the more usual case of `ndd.add(operand)` it would all be more straightforward, but this should handle the possible corner cases too.

Reply:
Gah, that comment was supposed to apply to the lines below, not above.